### ISSUE TITLE
Attempting to access value sets from within functions throws an exception in the engine

### ISSUE THREAD

**Issue by frankadrian314159** [DBCG/cql_engine#609]
Either of the following code snippets including functions:
library Test version '1'
using FHIR version '4.0.1'
valueset "VS": '20785'
context Unfiltered
/*
define testC: test3("VS")
define function test3(vs System.ValueSet): [Procedure: vs]
*/
/*
define testD: test4()
define function test4(): [Procedure: "VS"]
*/
/*
define testE: test5("VS")
define function test5(vs System.ValueSet): Code{code: 'A', system: 'foo'} in vs
*/

throws an exception in the engine with the following stack trace:

org.opencds.cqf.cql.engine.exception.CqlException: java.lang.IllegalArgumentException: Unable to locate ValueSet 20785
        at org.opencds.cqf.cql.engine.elm.execution.Executable.evaluate(Executable.java:37)
        at org.opencds.cqf.cql.engine.elm.execution.FunctionRefEvaluator.internalEvaluate(FunctionRefEvaluator.java:33)
        at org.opencds.cqf.cql.engine.elm.execution.Executable.evaluate(Executable.java:18)
        at org.opencds.cqf.cql.engine.elm.execution.ExpressionDefEvaluator.internalEvaluate(ExpressionDefEvaluator.java:19)
        at org.opencds.cqf.cql.engine.elm.execution.Executable.evaluate(Executable.java:18)
        at org.opencds.cqf.cql.engine.execution.CqlEngine.evaluateExpressions(CqlEngine.java:186)
        at org.opencds.cqf.cql.engine.execution.CqlEngine.evaluate(CqlEngine.java:164)
        at org.opencds.cqf.cql.evaluator.CqlEvaluator.evaluate(CqlEvaluator.java:89)
        at org.opencds.cqf.cql.evaluator.CqlEvaluator.evaluate(CqlEvaluator.java:76)
        at org.opencds.cqf.cql.evaluator.cli.command.CqlCommand.call(CqlCommand.java:159)
        at org.opencds.cqf.cql.evaluator.cli.command.CqlCommand.call(CqlCommand.java:33)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
        at picocli.CommandLine.access$1300(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
        at picocli.CommandLine.execute(CommandLine.java:2078)
        at org.opencds.cqf.cql.evaluator.cli.Main.run(Main.java:19)
        at org.opencds.cqf.cql.evaluator.cli.Main.main(Main.java:12)
Caused by: java.lang.IllegalArgumentException: Unable to locate ValueSet 20785
        at org.opencds.cqf.cql.evaluator.engine.terminology.BundleTerminologyProvider.expand(BundleTerminologyProvider.java:87)
        at org.opencds.cqf.cql.evaluator.engine.terminology.PrivateCachingTerminologyProviderDecorator.expand(PrivateCachingTerminologyProviderDecorator.java:47)
        at org.opencds.cqf.cql.evaluator.engine.terminology.PrivateCachingTerminologyProviderDecorator.in(PrivateCachingTerminologyProviderDecorator.java:25)
        at org.opencds.cqf.cql.evaluator.engine.retrieve.BundleRetrieveProvider.anyCodeInValueSet(BundleRetrieveProvider.java:89)
        at org.opencds.cqf.cql.evaluator.engine.retrieve.BundleRetrieveProvider.filterByTerminology(BundleRetrieveProvider.java:151)
        at org.opencds.cqf.cql.evaluator.engine.retrieve.BundleRetrieveProvider.retrieve(BundleRetrieveProvider.java:56)
        at org.opencds.cqf.cql.evaluator.engine.retrieve.PriorityRetrieveProvider.retrieve(PriorityRetrieveProvider.java:28)
        at org.opencds.cqf.cql.engine.data.CompositeDataProvider.retrieve(CompositeDataProvider.java:96)
        at org.opencds.cqf.cql.engine.elm.execution.RetrieveEvaluator.internalEvaluate(RetrieveEvaluator.java:57)
        at org.opencds.cqf.cql.engine.elm.execution.Executable.evaluate(Executable.java:18)
        ... 19 more

Basically, it looks as though whenever an InValueSet evaluation is done against a value set which has been passed to a function as a parameter, the terminology provider cannot find the target value set.

**Comment by c-schuler** [comment]
Interesting use case. 

I was able to successfully run the following in VSCode:

```
valueset "Marital Status": 'http://hl7.org/fhir/ValueSet/marital-status'

define "Call VSTestIn function":
    VSTestIn("Marital Status")

define function VSTestIn(vs System.ValueSet):
    Code { code: 'L', system: 'http://terminology.hl7.org/CodeSystem/v3-MaritalStatus' } in vs
```

Not sure how you're testing this, but is the ValueSet available in the terminology server/provider?

**Comment by frankadrian314159** [comment]
Yes, the ValueSet is available - it's being returned as the result of a
previous test, which I removed to shorten the test file. I am running the
code in the jar of cql-evaluator which I built from scratch yesterday.  I
heard that you folks had released a new version of the engine this morning.
Could that have fixed the issue? I am wondering if it's something as simple
as a versioning issue now.

Frank Adrian

**Comment by frankadrian314159** [comment]
Given that the exception is being thrown in the BundleTerminologyProvider, which is part of the evaluator, not the engine, I will shift my queries over to that component. You may close this issue.

**Comment by frankadrian314159** [comment]
In case you wanted to see it, I've attached the entire test file...

Frank Adrian
